[WIP] Spells/Death Grip fix. The pull effect shouldn't work on some bosses

chaodhib opened the pull request:
Fixes https://github.com/TrinityCore/TrinityCore/issues/14870
- An extra flag CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP has been created to handle the immunity to the pull component of the spell Death Grip
- a spell script has been written to the third spell called by the Death Grip chain in order to send to the client the message "immune" if the creature has the flag mentioned above.
- the 3 spell scripts are grouped together in the spell_dk for convenience.

Patchwerk (16028) is taken as the dummy:
![wowscrnshot_082615_012611](https://cloud.githubusercontent.com/assets/6612710/9482281/c47f795a-4b92-11e5-95e8-e3a9f6b6778d.jpg)
This screenshot is taken after the fix is applied. It show that indeed the taunt is applied but not the pull.

For some reason, the dk didnt got the same threat as the highest in list and only got aggro for the 3s duration of the taunt. Unsure if this is linked to the PR.

To do:
- [X] implement the extra flag
- [X] Make use of it to implement the immunity
- [x] Message "immune" should be shown to the caster if the mob is immune to the pull effect.
- [ ] Investigate the thing about the bugged threat. See if it's linked to this PR.
- [ ] Update the flag of each boss. And then, update the known exceptions. (SQL queries)

DDuarte commented on sql/updates/world/2015_XX_YY_ZZ_world.sql:
```diff
@@ -0,0 +1,2 @@
+UPDATE `world`.`creature_template` SET `flags_extra`='513' WHERE `entry`='16028';
+INSERT INTO `world`.`spell_script_names` (`spell_id`, `ScriptName`) VALUES ('49575', 'spell_dk_death_grip_pull');
```
- Remove hardcoded DB names
- Remove ' around numbers
- Add a DELETE FROM to delete the spell that you inserting.

DDuarte commented on src/server/scripts/Commands/cs_npc.cpp:
```diff
@@ -162,7 +162,8 @@ EnumName<CreatureFlagsExtra> const flagsExtra[FLAGS_EXTRA_COUNT] =
     CREATE_NAMED_ENUM(CREATURE_FLAG_EXTRA_ALL_DIMINISH),
-    CREATE_NAMED_ENUM(CREATURE_FLAG_EXTRA_DUNGEON_BOSS)
+    CREATE_NAMED_ENUM(CREATURE_FLAG_EXTRA_DUNGEON_BOSS),
+    CREATE_NAMED_ENUM(CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
```
Shouldn't this follow the same order of the enum above?

chaodhib replied:
Hm, I have no idea. Should it?

Shauren replied:
It doens't strictly have to but it will cause this flag to be printed last in chat if there are more flags set.

DDuarte commented on src/server/scripts/Spells/spell_dk.cpp:
```diff
@@ -1689,6 +1655,73 @@ class spell_dk_death_grip_initial : public SpellScriptLoader
 };
 
+// 49560 - Death Grip (second part)
+class spell_dk_death_grip : public SpellScriptLoader
```
Do you have to move this spell script? Why not adding the new one after the existing one?

chaodhib replied:
Because now, just for Death Grip, there is 3 spell scripts. For convenience, I thought it would be better to put them one after the other in the order they are called

DDuarte replied:
Fair enough, I didn't notice the 3rd spell.

joschiwald replied:
while your are moving spellscripts, move all 3 spells to previous location to keep alphabetical order

jackpoz replied:
moving spells around is quite messy when reviewing PRs

DDuarte commented on src/server/scripts/Spells/spell_dk.cpp:
```diff
@@ -1689,6 +1655,73 @@ class spell_dk_death_grip_initial : public SpellScriptLoader
+        SpellCastResult CheckCast()
+        {
+            Creature* creatureTarget = GetCaster()->ToCreature(); // the caster is the unit affected by Death Grip (see spell script of previous spell)
+            if (creatureTarget && creatureTarget->GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
```
Shouldn't this be checked in a more generic place and not in a very specific spellscript?
Doing it this way means that all spells with this mechanic need to be scripted individually. The flag should be checked for any spell with the pull effect.

ccrs replied:
Should be checked inside this:

````c++
void Spell::EffectJumpDest(SpellEffIndex effIndex)
{
    if (effectHandleMode != SPELL_EFFECT_HANDLE_LAUNCH)
        return;

    if (m_caster->IsInFlight())
        return;

    if (!m_targets.HasDst())
        return;

    // Init dest coordinates
    float x, y, z;
    destTarget->GetPosition(x, y, z);

    float speedXY, speedZ;
    CalculateJumpSpeeds(effIndex, m_caster->GetExactDist2d(x, y), speedXY, speedZ);
    m_caster->GetMotionMaster()->MoveJump(x, y, z, speedXY, speedZ);
}
````

chaodhib replied:
@ccrs I want the caster to be notified that the target is immuned to one of the effects of the spell (in this case, the pull component). When this method is reached, it's already too late to send that "immune" message.

@DDuarte Sure but I don't know of any other spell that could make use of that flag actually. Do you have any in mind?

DDuarte replied:
> his own ability wont work

If target is caster, ignore? Your scenario seems highly unlikely

ccrs commented:
quoting the great @Shauren Never delete by spell_id, always on scriptnames

robinsch commented:
About the immune message: https://github.com/TrinityCore/TrinityCore/issues/14968#issuecomment-118596838 and below

Warpten commented:
Jesus the wall of text.

chaodhib commented:
@robinsch Yep, I'm working with that as well.

Apparently, I confused SpellCastResult and SpellMissInfo the same way i confused SendCastResult() and SendSpellMiss(). The code checking for the flag shouldn't be in a spellscript's CheckCast(), otherwise the wrong message is sent.

chaodhib commented:
Quick and easy solution:
![wowscrnshot_082815_174912](https://cloud.githubusercontent.com/assets/6612710/9550770/9cf36af0-4dad-11e5-97d4-42dadc7ead47.jpg)

```C++
void Spell::EffectJumpDest(SpellEffIndex effIndex)
{
    if (effectHandleMode != SPELL_EFFECT_HANDLE_LAUNCH)
        return;

    if (m_caster->IsInFlight())
        return;

    if (!m_targets.HasDst())
        return;

    if (m_caster != m_originalCaster)
    {
        Creature* casterCreature = m_caster->ToCreature();
        if (casterCreature && casterCreature->GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
        {
            m_originalCaster->SendSpellMiss(m_caster, this->GetSpellInfo()->Id, SpellMissInfo::SPELL_MISS_IMMUNE);
            return;
        }
    }

    // Init dest coordinates
    float x, y, z;
    destTarget->GetPosition(x, y, z);

    float speedXY, speedZ;
    CalculateJumpSpeeds(effIndex, m_caster->GetExactDist2d(x, y), speedXY, speedZ);
    m_caster->GetMotionMaster()->MoveJump(x, y, z, speedXY, speedZ);
}
```
* Is generic.
* Send the correct "immune" message
* the taunt is still applied
* but... in the spell effect handler, use of the method SendSpellMiss() which I guess is forbidden?

jackpoz commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -1045,6 +1045,16 @@ void Spell::EffectJumpDest(SpellEffIndex effIndex)
+        Creature* casterCreature = m_caster->ToCreature();
+        if (casterCreature && casterCreature->GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
+        {
+            m_originalCaster->SendSpellMiss(m_caster, this->GetSpellInfo()->Id, SpellMissInfo::SPELL_MISS_IMMUNE);
```
please avoid using "this->" when it's unneeded

jackpoz commented on src/server/scripts/Spells/spell_dk.cpp:
```diff
@@ -639,38 +639,81 @@ class spell_dk_death_gate : public SpellScriptLoader
+            // Patch 3.3.3 (2010-03-23): Minimum range has been changed to 8 yards in PvP.
+            Unit* target = GetExplTargetUnit();
+            if (target && target->GetTypeId() == TYPEID_PLAYER)
+                if (caster->GetDistance(target) < 8.f)
```
use IsInRange() for range checks

chaodhib replied:
I rather not touch this since I didn't wrote it and I don't know the details of implementation for these methods and it has nothing to do with the issue I'm trying to solve.

chaodhib commented:
I'm thinking about dropping the third point (sending an immune message) from this PR. The core doesn't support it very well. 

A SpellMissInfo::SPELL_MISS_IMMUNE should be sent each time the target is immune to one of the spell effects. This functionally should have its own PR. What do you guys think?

ccrs commented:
AFAIK on retail (right now) no inmune message is sent when a DK Death Grip a Boss.

chaodhib commented:
@ccrs a boss who is immune to pull?

Because I found a 4.x video where an immune is shown if the boss is immune to the pull effect (cf https://github.com/TrinityCore/TrinityCore/issues/14870#issuecomment-134713447)

ccrs commented:
I mean the inmune message. I dont see it on your example https://youtu.be/7cALSvQFoFI?t=1m46s

ccrs commented:
I guess ur right

chaodhib commented:
![7005bac4-4b6a-11e5-8579-7d07d42c04d8](https://cloud.githubusercontent.com/assets/6612710/9705393/524ccd30-54c3-11e5-8ccb-2d5979e6e82f.jpg)

Rushor commented:
btw did you got any strange aggrobehaviour while testing the creatures? 
for some reason creatures with this new flag no longer chase the victim :/

chaodhib commented:
Yep: https://github.com/TrinityCore/TrinityCore/issues/15403

Rushor commented:
moving the changes you added in void Spell::EffectJumpDest(SpellEffIndex effIndex) to the spellscript of deathgrip solves the strange issue, that creatures no longer follow the victim

jackpoz commented:
what's the current status of this PR ?

ccrs commented:
If there is no objection to this:

````diff
     if (!m_targets.HasDst())
          return;
  
 +    if (m_caster != m_originalCaster)
 +    {
 +        Creature* casterCreature = m_caster->ToCreature();
 +        if (casterCreature && casterCreature->GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
 +        {
 +            m_originalCaster->SendSpellMiss(m_caster, GetSpellInfo()->Id, SpellMissInfo::SPELL_MISS_IMMUNE);
 +            return;
 +        }
 +    }
 +
      // Init dest coordinates
      float x, y, z;
      destTarget->GetPosition(x, y, z);
````
(send SendSpellMiss when effect is handled)

Its just a matter of inserting creature entries inside the sql

chaodhib commented:
The status is that I proposed a (unfinished) solution and I'm waiting for confirmation that it's an acceptable one. I fear that m_originalCaster->SendSpellMiss() shouldn't be called from that method due to a design choice. If I'm right and this solution isn't acceptable, then the spell system will have to be modified, which will be a bit more tricky.

Either way, I'm waiting for feedback before I resume working on this issue.

joschiwald commented:
i have 2 possible solutions:

first one: set mechanic of the initial grip spell to none, then you can control taunt and grip separatly by mechanic_immune_mask (no new flag required)

second one:
add a block to Creature::IsImmunedToSpellEffect that handles your CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP, if has flag && effect->Mechanic == MECHANIC_GRIP then return true;

ccrs commented:
I vote for the second solution :P

chaodhib commented:
The 2nd solution doesn't work. Putting in Creature::IsImmunedToSpellEffect():
* ```if (spellInfo->Mechanic == MECHANIC_GRIP && GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
        return true;```
wont work because the first spell has Mechanic set to MECHANIC_GRIP (so the first spell is entirely blocked and the second spell is never launched).
* ```if (spellInfo->GetEffectMechanic(index) == MECHANIC_GRIP && GetCreatureTemplate()->flags_extra & CREATURE_FLAG_EXTRA_NO_PULL_BY_GRIP)
        return true;```
wont work because for some reason (a mistake possibly), GetEffectMechanic() returns the Mechanic of the spell if it doesnt find a Mechanic to the spell effect. And then, the same thing as above happens.

The 1st solution works but setting the mechanic to none has to be done for 2 spells: the initial one (49576) and the main one (49560). I'm not really fond of this solution since there is no "immune" message thrown back to the player.

I've been looking for a solution that returns that "immune" message but to no avail so far.

chaodhib commented:
Do we have sniff on Death Grip? Are we sure it's the spells 49576, 49560 and 49575 that are used?

chaodhib commented:
Finally made some progress. I was about to give up but I did what I should have done wayyyy sooner: I looked at sniffs (stupid stupid me). Anyway, I discovered that the spell 51399 is also involved in the Death Grip spell chain.

Here is the result using SniffExplorer when I look for every SPELL_GO and SPELL_START messages with the spells 49560, 49576, 49575 and 51399 : 
* sniff on naxx
https://gist.github.com/chaodhib/34d0fe967958c996382d#file-death-grip-analysis-txt
* sniff on uld
https://gist.github.com/chaodhib/2b5d9f295d0605f7d461

* Also, here is a detail for a Death Grip cast against Patchwerk (immune to pull effect, not immune to taunt). it's in the first sniff:
https://gist.github.com/chaodhib/4123d3c31e2912d5bc6f#file-death-grip-on-patchwerk-txt
* and the detail for a Death Grip cast against a player (immune to taunt, not immune to the pull). It's in the end of the first sniff. The dk was dueling a druid:
https://gist.github.com/chaodhib/5cf55e6e91928baf4e25#file-death-grip-on-player-txt

I'm gonna do further analysis when I find time. What I can already say is that 49576 is indeed the first spell and it's cast all the time. Now depending on a certain condition which I'm not sure (my guess would be it checks for a flag in the Creature Template), either only 51399 (taunt only) is launched. or 49560 and 49575 (taunt + pull effect).

tkrokli commented:
Nice progress in getting proper background info. I hope this will help you with the PR development. :+1:

Trisjdc commented:
Why not just use immunity to MECHANIC_GRIP? No extra hacky flags, just good old reliable `creature_template`.`mechanic_immune_mask`.

ccrs commented:
that also makes the npc inmune to taunt afaik

Treeston commented:
Do we have a spellscript hook for "effect was immuned by target"? I feel like this would be the solution.

If Effect 0 of 49576 hits target: Trigger spell 49560 at target.
If Effect 0 of 49576 is immuned by target: Trigger spell 51399 at target.

PS: Not getting top threat is blizzlike, it's "only" a force attack debuff, not a "true" taunt.

Treeston commented:
If we don't have that spell script, we should add that spell script.

chaodhib commented:
I cleaned the git gists above. Basically, from sniff, we can observe 3 typical situations:

* Death Grip against Ignis  (immune to pull effect, not immune to taunt):

The spell 49576 (Effect 0 SPELL_EFFECT_DUMMY) is launched. It launches the spell 51399 but also **throws an immune**.
The spell 51399 (Effect 0 AURA_MOD_TAUNT & Effect 1 SPELL_EFFECT_ATTACK_ME) **hits**.

* Death Grip against a player (not immune to the pull effect, immune to the taunt):

The spell 49576 is launched. It **hits** and launches 49560.
The spell 49560 (Effect 0 SPELL_EFFECT_DUMMY, Effect 1 & 2 AURA_MOD_TAUNT & SPELL_EFFECT_ATTACK_ME) **throws an immune** but it makes the target launch the spell 49575.
The spell 49575 (Effect 0 SPELL_EFFECT_JUMP_DEST) is launched by the target (and cannot miss since it's a self spell).

* Death Grip against a "regular" mob (not immune to both):

The spell 49576 is launched. It **hits** and launches 49560.
The spell 49560 **hits** and makes the target launch 49575.
The spell 49575 **hits**.

Treeston commented:
So what I described above?

> Do we have a spellscript hook for "effect was immuned by target"? I feel like this would be the solution.

>If Effect 0 of 49576 hits target: Trigger spell 49560 at target.
>If Effect 0 of 49576 is immuned by target: Trigger spell 51399 at target.

>PS: Not getting top threat is blizzlike, it's "only" a force attack debuff, not a "true" taunt.

tkrokli commented:
The branch `chaodhib:death_grip_fix` is 894 commits behind `TrinityCore:3.3.5`. I don't know if there is a shortcut to fix the unresolved conflicts, I would think a rebase is required (I may be wrong, of course).

Treeston commented:
Yeah, rebase when you have time please.

chaodhib committed:
Making most bosses immune to the pull effect of death grip.

Conflicts:
	src/server/game/Entities/Creature/Creature.h
	src/server/game/Spells/SpellEffects.cpp
	src/server/scripts/Commands/cs_npc.cpp

chaodhib committed:
Using the other death grip spell, 51399, in case the target is perma immune to the pull effect.

chaodhib commented:
I rebased and implemented what I described above, expect no "immune" is shown.

Treeston commented:
Targets should use `MECHANIC_GRIP` immunity instead of the new extra flag. The "taunt only" DG spell does not have `MECHANIC_GRIP` -it is, in fact, the only spell of the chain without this mechanic, and I am guessing this is why.

If the creature is immune to the initial spell (`MECHANIC_GRIP` immunity), then the "taunt only" DG spell should fire, with the first spell providing the initial "Immune" message. If the creature is immune to both the grip and the taunt, then "Immune" will show from the first spell and then from the second ("taunt only") spell.

chaodhib commented:
@Treeston 
Interesting. I didnt understood your "trigger" thing until now. Your solution would solve the first "immune" message. Not the second however. How explain the "immune" message showing up when the target is immune to the taunt?

> The spell 49560 (Effect 0 SPELL_EFFECT_DUMMY, Effect 1 & 2 AURA_MOD_TAUNT & SPELL_EFFECT_ATTACK_ME) throws an immune but it makes the target launch the spell 49575.

Unit::SpellHitResult() calls Unit::IsImmunedToSpell().
IsImmunedToSpell() will call IsImmunedToSpellEffect() for spell effects 0, 1 and 2. IsImmunedToSpellEffect() will returns true for spell effects 1 and 2. false for spell effect 0.
Unit::IsImmunedToSpell() will then return false.
And finally, Unit::SpellHitResult will NOT return SpellMissInfo::SPELL_MISS_IMMUNE

Treeston commented:
That sounds like a core bug, then.

chaodhib commented:
Yeah.

I see 2 solutions:
* Introduce a new kind of spell script hook which allows us to override the SpellHitResult of each units hit by a spell.
* If a unit hit by a spell is immune to at least one of the spell effects (instead of all), a "immune" should be sent. This needs data because I wasn't able to confirm/infirm that this is blizz-like.

Treeston commented:
Anecdotally, the second one is blizzlike.

chaodhib commented:
What do you mean? Do you have proof?

Treeston commented:
[First reference to it I could find.](http://wow.allakhazam.com/forum.html?forum=243&mid=12399324443464653) Ctrl-F for "immune".

Treeston commented:
[Another reference](http://us.battle.net/wow/en/forum/topic/17347645772) for Frostbolt doing the same thing.

chaodhib commented:
@Treeston 
To be honest, this is a bit thin. Some evidence by sniff would be great.

mik1893 commented:
i'm quite sure from old good playtimes that you get an immune (as @Treeston is saying).

Aokromes commented:
Any reason to don't merge this?

chaodhib commented:
It's not ready. The implementation is not good.

Treeston commented:
Please don't remove the WIP tag unless you are the author.

chaodhib commented:
I'm not really working much on this PR anymore (I'm focusing on parts of the game). So if someone wants to work on this, I can give access to my repo.

Treeston commented:
The two things mentioned in https://github.com/TrinityCore/TrinityCore/pull/15354#issuecomment-182638837 still need to be added before this can be continued properly.

Aokromes commented:
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
Conflicting files
src/server/game/Entities/Creature/Creature.h
src/server/scripts/Commands/cs_npc.cpp

chaodhib commented:
It is pointless of me to keep this PR open. I won't work on this again (focusing on other parts of the core). If someone wish to fix the DG issue, better to open a new PR.
